### ISSUE TITLE
Restore app window after user closes with red button

### ISSUE THREAD

**Issue by Montoya** [MacGapProject/MacGap2#56]
Hi, so when the user clicks the red button in the toolbar on the top left, the app window closes. 

I think it's really cool that the user can then click on the dock icon to restore the app window just as it was. Now I'm trying to add a menu item that does the same thing. I've already created the menu item in MainMenu.xib, I just need to give it a callback. So far I have: 

```
MacGap.Menu.getItem("File").submenu.getItem('New Window').callback = function() { 
    /* ... */
}; 
```

I've tried a few things and nothing worked. What should be in the callback?

**Comment by Montoya** [comment]
bump!

**Comment by jeff-h** [comment]
This is a bit trickier than it may appear, since MG2 is (now^) capable of showing multiple windows, each with a different URL. I imagine that when you close the window and click the dock icon, it's actually opening a new window with the default URL as specified in `config.json`. That's quite distinct from opening the old window just as it was.

This means your JS app is going to think it's just been started, and will need to reinstate whatever it was doing last. You'd need to handle this on the JS side... not a trivial exercise if your app is complex, I imagine.

Anyway, now that Window.open(url) is available, you should be able to implement that inside your menu callback.

^ I just pushed a simpler Window.open() method and documented it at http://docs.macgap.com/ref/window.html

**Comment by Montoya** [comment]
> I imagine that when you close the window and click the dock icon, it's actually opening a new window with the default URL as specified in config.json.

No, you should try it. It opens the window in the same state that you left it when you closed it (by clicking the red button).

**Comment by jeff-h** [comment]
You're right, sorry. I read the code wrongly...

I've added a new method called `reopenFirst()` which reopens the first-opened window.

Call it like so:

```js
MacGap.Window.reopenFirst();
```

It feels clumsy to me in all respects — philosophically (we're avoiding more fundamental multiple-document architecture issues), the method name, the code location...

Thoughts?

**Comment by Montoya** [comment]
OK well, 1st thought is, it works for me so I love it! Second thought is, even in a multi-window app, I assume that the first window is going to be the main entry point for the app, so if the user has closed all the windows then they probably want to reopen that first window if they are trying to reopen the app. 3rd thought is, this almost seems like a core MacGap method that belongs up top, like MacGap.activate()

Hope that helps. 

Also I ran into a couple issues while testing this. The first is, I can't get event listeners working for MacGap events. As a simple test I have this inside my $(document).ready: 

```
document.addEventListener('resize', function() { 
    alert("WHAT"); 
}, true); 
```

I can resize the app window by dragging the bottom left corner and this callback does not seem to be fired. 

Also the method for closed in window.m doesn't even trigger an event, as far as I can tell: 

```
- (void) windowClosed: (NSNotification*)notification
{
    self.webView = nil;
    self.windowController = nil;
}
```

??

**Comment by rawcreative** [comment]
I haven’t had time to look at any of the code changes jeff mentioned previously, but to address the issues you mentioned:

1. You’re attaching the event listeners to the wrong object, all window events are fired on the window object, not the document.
2. The windowClosed method doesn’t fire an event. It would be pointless to, at least the way things are currently because each window has it’s own javascript context that’s tied to the web view. When the window is closed the web view and it’s context is released, along with any event listeners you would’ve attached.

**Comment by Montoya** [comment]
OK cool. 

1. You are correct, and also the documentation should mention this! 
2. Well, it does if you fire the event before closing the window: 

```
- (void) windowClosed: (NSNotification*)notification
{
    [self triggerEvent:@"closed"];
    self.webView = nil;
    self.windowController = nil;
}
```

This way I can listen for the event and set my "Reopen" menu item to enabled when the window is closed.

**Comment by Montoya** [comment]
Also, call me crazy, but I just did what I should have done a long time ago: disabled the close button in MainWindow.xib. Since my app is a single-window app, this makes sense :)

**Comment by jeff-h** [comment]
@rawcreative please correct me if I'm wrong, but adding `[self triggerEvent:@"closed"];` to the top of `windowClosed()` won't work, as any event handlers on the JS side won't predictably have sufficient time to do stuff before the webview and windowController are released. You might be lucky most of the time if you're doing something very fast on the JS side, but I don't think you could rely on it.
